perfkitbenchmarker/providers/gcp/gcp_dpb_dataflow.py

```diff
@@ -152,4 +186,303 @@ def GetMetadata(self):
     basic_data = super(GcpDpbDataflow, self).GetMetadata()
     basic_data['dpb_dataflow_runner'] = FLAGS.dpb_dataflow_runner
     basic_data['dpb_dataflow_sdk'] = FLAGS.dpb_dataflow_sdk
+    basic_data['dpb_job_id'] = self.job_id
     return basic_data
+
+  def GetJobStatus(self):
+    cmd = util.GcloudCommand(self, 'dataflow', 'jobs', 'show', self.job_id)
+    cmd.flags = {
+        'project': self.project,
+        'format': 'json',
+    }
+
+  def GetStats(self):
+    """Collect series of relevant performance and cost stats."""
+    stats = {}
+    stats['total_vcpu_time'] = self.GetMetricValue('TotalVcpuTime')/3600         # vCPU-hr
+    stats['total_mem_usage'] = self.GetMetricValue('TotalMemoryUsage')/1024/3600 # GB-hr
+    stats['total_pd_usage'] = self.GetMetricValue('TotalPdUsage')/3600           # GB-hr
+    # TODO(rarsan): retrieve BillableShuffleDataProcessed
+    # and/or BillableStreamingDataProcessed when applicable
+    self.job_stats = stats
+    return stats
+
+  def CalculateCost(self, type=DATAFLOW_TYPE_BATCH):
+    if type not in (DATAFLOW_TYPE_BATCH, DATAFLOW_TYPE_STREAMING):
+      raise ValueError(f'Invalid type provided to CalculateCost(): {type}')
+
+    if not self.job_stats:
+      self.GetStats()
+
+    total_vcpu_time = self.job_stats['total_vcpu_time']
+    total_mem_usage = self.job_stats['total_mem_usage']
+    total_pd_usage = self.job_stats['total_pd_usage']
+
+    cost = 0
+    if type == DATAFLOW_TYPE_BATCH:
+      cost +=  total_vcpu_time * VCPU_PER_HR_BATCH
+      cost +=  total_mem_usage * MEM_PER_GB_HR_BATCH
+    else:
+      cost += total_vcpu_time * VCPU_PER_HR_STREAMING
+      cost += total_mem_usage * MEM_PER_GB_HR_STREAMING
+    
+    cost += total_pd_usage * PD_PER_GB_HR
+    # TODO(rarsan): Add cost related to per-GB data processed by Dataflow Shuffle
+    # (for batch) or Streaming Engine (for streaming) when applicable
+    return cost
+
+  def _PullJobMetrics(self, force_refresh=False):
+    """Retrieve and cache all job metrics from Dataflow API"""
+    # Skip if job metrics is already populated unless force_refresh is True
+    if self.job_metrics is not None and not force_refresh:
```

Comment:
Also return if self.job_id is None

Reply:
Should we raise an exception if job_id is None?
When self.job_id is None (which is not expected), self.job_metrics will be None, and then `GetMetricValue` will error out. I'm tempted to fail fast, rather than having a sequence of job runs proceed without any stats collected. What do you think?

Reply:
I think that failing on job_id is None would make the code simpler so I support the decision, I am just worried that parsing it from a stderr human readable message is kinda brittle, isn't there a better way to do it?

Reply:
Sounds good. Updated.
I agree job_id parsing is bit brittle since parent dataflow provider originally relies on java executable. For the subclass'ed dataflow template provider, it's more robust as job_is is returned as part of stdout response (JSON object) by gcloud command.